### PR TITLE
Soulcrypt Addition

### PR DESCRIPTION
**Pull request by Andromeda-K22**
## About The Pull Request

**In the interest of transparency, I must state that this was a code bounty offered by Nestor.**

The **Soulcrypt** is a replacement of the current Neotheology system for revival - and an extensible system for various useful features, somewhat like the Nanite Implant Framework from Virgo.

The system is defined as needing to do a few things, listed here:
- Hold a ref to the host's mind datum.
- Hold a copy of their DNA.
- Function like an ID card, with access that can be modified at will by someone with sufficient access. 
- Have an extensible module system to allow for the easy implementation of new player-facing functions.
- Have wear and energy usage mechanics, meaning usage of the soulcrypt's module functions will require an active upkeep in terms of food and eventually, a visit to medical for routine maintenance. 
- Provide information relating to the death of the host; it will directly tell them how much they can remember when they die, and do so again when they are cloned. This varies from 30 seconds before death normally, to 5 minutes before death if it has been subjected to an electromagnetic pulse. 

## Changelog
:cl:
add: Soulcrypt Base Implant
add: Soulcrypt Module System
/:cl:

### CODE REVIEW

File: code/modules/soulcrypt/module.dm [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	owner.send_host_message(activation_msg, MESSAGE_NOTICE)
+	if(!nanomodule)
+		nanomodule = new nanomodule_path
+		nanomodule.host = src
+	ui_interact(user)
+
+/datum/soulcrypt_module/proc/deactivate(var/force_close = FALSE)
+	active = FALSE
+	var/deactivation_msg = "<b>[name]:</b> [deactivation_message]"
+	if(nanomodule)
```
Comment (PsiOmegaDelta) [inline code/modules/soulcrypt/module.dm:69]: To QDEL_NULL macro already makes this check

File: code/modules/soulcrypt/module.dm [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	var/list/req_access = list()
+	var/list/req_one_access = list()
+
+	var/obj/item/weapon/implant/soulcrypt/owner //The soulcrypt that owns us.
+
+	var/datum/nano_module/soulcrypt/nanomodule //Our nanomodule, used for proper interaction and such.
+	var/nanomodule_path
+
+/datum/soulcrypt_module/Destroy()
+	if(nanomodule)
```
Comment (PsiOmegaDelta) [inline code/modules/soulcrypt/module.dm:30]: The QDEL_NULL macro already checks this

**Comment by bearrrrrrrr** [comment]
can you please add software so that you can send pics fof your penis

(no but a cameras could be cool idk lol)

also i'd raise the time on what you can remember on the remember threshhold, 30 sec is a tad low

**Comment by Andromeda-K22** [comment]
I'l see what I can do about cameras.

I was directed to make the threshold 30 seconds by Nestor.

**Comment by EvilJackCarver** [comment]
@bearrrrrrrr ease back on the NSFW a tad, please

**Comment by Andromeda-K22** [comment]
# Todo
Add remote access management computer/tool of some kind for soulcrypt access
Add way to actually install new modules into the soulcrypt

roughly 80% feature complete at this moment in time.

**Comment by Andromeda-K22** [comment]
nearing completion and thus review; last thing to do is add soulcrypts to the loadout.

**Comment by Andromeda-K22** [comment]
Ready for review.

**Comment by EvilJackCarver** [comment]
I'll review and pull the don't-merge flag here as soon as I finish testing Nestor's hotfix.

File: code/modules/soulcrypt/module.dm [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	if(owner.wearer.incapacitated())
+		return FALSE
+
+	return TRUE
+
+/datum/soulcrypt_module/proc/activate(var/mob/user)
+	if(!check_can_activate())
+		return
+	active = TRUE
+	var/activation_msg = "<b>[name]:</b> [activation_message]"
```
Comment (EvilJackCarver) [inline code/modules/soulcrypt/module.dm:60]: Ambiguous, but acceptable. Consider prefixing the var with an underscore (`_activation_msg`) to make it a bit clearer that it's a local var (also a few lines below).

File: code/modules/soulcrypt/statpanel.dm [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+			if(crypt.integrity < 20)
+				crypt_status = "Service Needed Soon"
+			else
+				crypt_status = "Operating Normally"
+		if(SOULCRYPT_LOW_POWER)
+			crypt_status = "Insufficient Energy!"
+		if(SOULCRYPT_DISRUPTED)
+			crypt_status = "System Failure!"
+		else
+			crypt_status = "Unknown - Error"
```
Comment (EvilJackCarver) [inline code/modules/soulcrypt/statpanel.dm:26]: May be a good idea to use a `throw EXCEPTION()` call here if a sanity check fails; that way it'll generate a stack trace and we'll be able to debug it easier.

**Comment by Andromeda-K22** [comment]
Some more stuff incoming; don't merge yet.

**Pull request merged**